docs/ShapeInference.md

```diff
@@ -0,0 +1,62 @@
+# ONNX Shape Inference
+
+ONNX provides an optional implementation of shape inference on ONNX
+graphs. This implementation covers each of the core operators, as well
+as provides an interface for extensibility. Therefore, you may choose
+to invoke the existing shape inference functionality on your graphs,
+or to define shape inference implementations to go along with your
+custom operators (or both!).
+
+## Invoking Shape Inference
+
+Shape inference can be invoked either via C++ or Python. The Python
+API is described, with example,
+[here](docs/PythonAPIOverview.md#running-shape-inference-on-an-onnx-model).
+
+The C++ API consists of a single function
+
+```
+shape_inference::InferShapes(
+    ModelProto& m,
+    const ISchemaRegistry* schema_registry);
```

Comment:
(unrelated to this PR) IMO this `schema_registry` is really an unfortunate complexity added to the API